python-checks/src/main/java/org/sonar/python/checks/EinopsSyntaxCheck.java

```diff
@@ -0,0 +1,169 @@
+  }
+
+  private static void checkEinopsSyntax(SubscriptionContext ctx) {
+    CallExpression callExpression = (CallExpression) ctx.syntaxNode();
+    Symbol calleeSymbol = callExpression.calleeSymbol();
+    if (calleeSymbol != null && calleeSymbol.fullyQualifiedName() != null && FQN_TO_CHECK.contains(calleeSymbol.fullyQualifiedName())) {
+      extractPatternFromCallExpr(callExpression).ifPresent(stringLiteral -> {
+        var maybePattern = toEinopsPattern(stringLiteral);
+        if (maybePattern.isPresent()) {
+          var pattern = maybePattern.get();
+          checkForEllipsisInParenthesis(ctx, pattern);
+          checkForUnbalancedParenthesis(ctx, pattern);
+          checkForUnusedParameter(ctx, callExpression.arguments(), pattern);
+        } else {
+          ctx.addIssue(callExpression.callee(), "Provide a valid einops pattern.");
+        }
+      });
+    }
+  }
+
+  private static void checkForUnusedParameter(SubscriptionContext ctx, List<Argument> arguments, EinopsPattern pattern) {
+    List<String> argsToCheck = arguments.stream()
+      .map(TreeUtils.toInstanceOfMapper(RegularArgument.class))
+      .filter(Objects::nonNull)
+      .filter(arg -> arg.expression().is(Tree.Kind.NUMERIC_LITERAL))
+      .filter(arg -> arg.keywordArgument() != null)
+      .map(arg -> arg.keywordArgument().name())
+      .filter(argName -> !pattern.lhs.identifiers.contains(argName) || !pattern.rhs.identifiers.contains(argName))
+      .toList();
+
+    if (!argsToCheck.isEmpty()) {
+      var isPlural = argsToCheck.size() > 1;
+      var missingParameters = argsToCheck.stream().collect(Collectors.joining(", "));
+      var missingParametersMessage = String.format("the parameter%s %s do%s not appear in the pattern", isPlural ? "s" : "", missingParameters, isPlural ? "" : "es");
+      ctx.addIssue(pattern.originalPattern(), String.format(MESSAGE_TEMPLATE, missingParametersMessage));
+    }
+  }
+
+  private static void checkForUnbalancedParenthesis(SubscriptionContext ctx, EinopsPattern pattern) {
+    pattern.lhs.state.errorMessage.or(() -> pattern.rhs.state.errorMessage)
+      .ifPresent(message -> ctx.addIssue(pattern.originalPattern(), String.format(MESSAGE_TEMPLATE, message)));
+  }
+
+  private record EinopsPattern(StringLiteral originalPattern, EinopsSide lhs, EinopsSide rhs) {
+  }
+
+  private record EinopsSide(String originalPattern, Set<String> identifiers, ParenthesisState state) {
+  }
+
+  private record ParenthesisState(boolean hasOpenParenthesis, Optional<String> errorMessage) {
+  }
+
+  private static Pattern ellipsisPattern = Pattern.compile("\\(.*\\.{3}.*\\)");
+
+  private static void checkForEllipsisInParenthesis(SubscriptionContext ctx, EinopsPattern pattern) {
+    if (ellipsisPattern.matcher(pattern.lhs.originalPattern).find()) {
+      ctx.addIssue(pattern.originalPattern(), String.format(MESSAGE_TEMPLATE, "Ellipsis inside parenthesis on the left side is not allowed"));
```

Comment:
Couldn't the lhs also contain the ellipsis unicode character?

Reply:
Yes you are right thanks!